### ISSUE TITLE
Make whitelistDomain check the domain is valid.

### ISSUE THREAD

**Issue by ancarda** [ancarda/security-headers#2]
A user on IRC asked for the following:

>I also feel like whitelistDomain on CSP needs to check the value I am trying to whitelist and possibly throw an Exception if it isn’t a valid thing.
>But I would have to recheck the CSP standard to see what is and isn’t allowed in there

This issue would cause a backwards compatibility break due to the introduced exception.

**Comment by peter279k** [comment]
Hi @ancarda, I want to help this features.

Can you present more details about implementing this feature?

Thanks.

**Comment by ancarda** [comment]
Hi @peter279k,

If you want to try implementing this, please go ahead. There's some functions in the CSP class like `withXXXFromDomain`. It would be nice if they could check if the domain looks valid. Some valid domains include:

* `localhost`
* `example.com`
* `xn--j1ay.xn--p1ai` (rendered `кц.рф`)
* `ancarda.blob.ancarda.net`
* `test.invalid`

If a domain isn't valid, a new exception like `Exception\InvalidDomainException` could be thrown.

I'm not 100% sure how to validate this as I don't fully know what constitutes an invalid domain. See if you can find:

1) A few invalid domains to be used for unit tests and
2) An RFC or some widely accepted regex for passing the above, but not the invalid tests.

Doesn't have to be perfect! Some validation is better than no validation IMO.

/cc @Zegnat might just be amazing and know a list of invalid domains.

**Comment by Zegnat** [comment]
As with all these things, matching exactly is super hard because there are just way too many rules spread over way too many RFCs. I’d recommend just having PHP handle it:

```php
if (false === filter_var($domain, FILTER_VALIDATE_DOMAIN)) {
    throw new Exception\InvalidDomainException();
}
```

Though I am not sure everything that makes for a valid domain is also a valid CSP value. I haven’t done much research on CSP these last couple of months, and am not currently planning to do another deep dive on it.

**Comment by peter279k** [comment]
I also agree with @Zegnat, because of many RFC rules that we cannot implement as well.

They're too difficult to implement this feature without using native PHP `filter_var` function.
